kore/src/Kore/Step/SMT/Representation/All.hs

```diff
@@ -46,9 +57,46 @@ sorts).
 build
     :: VerifiedModule Attribute.Symbol
     -> Map.Map Id Attribute.Constructors
-    -> AST.SmtDeclarations
+    -> Step.AST.SmtDeclarations
 build indexedModule sortConstructors =
-    resolve (sorts `AST.mergePreferFirst` symbols)
+    removeDuplicateConstructorDeclarations
+    $ resolve (sorts `Step.AST.mergePreferFirst` symbols)
   where
     sorts = Sorts.buildRepresentations indexedModule sortConstructors
     symbols = Symbols.buildRepresentations indexedModule
+
+removeDuplicateConstructorDeclarations
```

Comment:
I think it would be preferable to solve this in `kore/src/Kore/Step/SMT/Representation/Symbols.hs` when merging the different symbol types, i.e. in this code:
```
    listToDeclarations builtinDeclarations
    `AST.mergePreferFirst` listToDeclarations smtlibDeclarations
    `AST.mergePreferFirst` listToDeclarations constructorDeclarations
```
i.e. instead of `AST.mergePreferFirst` we could use something else that uses the constructor, or we could put the constructors before the smtlibDeclarations.

Also, note that, if you remove a symbol from this list, it likely means that you will not be able so send it to the SMT solver. Using the constructor entry for the symbol would solve this issue.

Could you add a test for this? A combination between `Constructors work` and `Uses smtlib name for constructor` in `kore/test/Test/Kore/Step/SMT/Symbols.hs`, but using an actual encoded predicate, would be nice.

I may approve this if the test works, even of you keep the current solution.

Reply:
@virgil-serbanuta I'm sorry but I haven't been able to implement a test which would reproduce the issue this PR intends to fix. I think this is because I haven't been able to figure out a sequence of sentence declarations which would reproduce this. Also, we didn't really get to the reason why there are duplicate symbol declarations, only that they are not resolved by the current master code. 

For example, when running one of the specs where this happened, the `SmtDeclarations` in `Kore.Step.Representation.All` would end up with the following declarations:
```
( Id {getId = "SortWordStack", idLocation = AstLocationFile (FileLocation {fileName = "test-sum-to-n-vdefinition.kore", line = 153, column = 8})}
, Sort { smtFromSortArgs, declaration = SortDeclarationDataType (DataTypeDeclaration {name = "|HB_SortWordStack|", typeArguments = [], constructors = [Constructor {name = "_dotWS", arguments = []},Constructor {name = "_WS_", arguments = [ConstructorArgument {name = "|HB_Lbl'UndsColnUndsUnds'EVM-TYPES'Unds'WordStack'Unds'Int'Unds'WordStack1|", argType = Atom "Int"},ConstructorArgument {name = "|HB_Lbl'UndsColnUndsUnds'EVM-TYPES'Unds'WordStack'Unds'Int'Unds'WordStack2|", argType = Atom "|HB_SortWordStack|"}]}]})}
)

( Id {getId = "Lbl'Stop'WordStack'Unds'EVM-TYPES'Unds'WordStack", idLocation = AstLocationFile (FileLocation {fileName = "test-sum-to-n-vdefinition.kore", line = 492, column = 10})}
, Symbol { smtFromSortArgs, declaration = SymbolDeclaredDirectly (FunctionDeclaration {name = "_dotWS", inputSorts = [], resultSort = Atom "|HB_SortWordStack|"})}
)
```
The first entry declares a sort with a constructor named `_dotWS`, and the second is a symbol declaration with the same name. The only sentence in the definition which has an `smtlib` attribute with this name is the following:
```
  symbol Lbl'Stop'WordStack'Unds'EVM-TYPES'Unds'WordStack{}() : SortWordStack{} [functional{}(), constructor{}(), org'Stop'kframework'Stop'attributes'Stop'Source{}("Source(/home/ttuegel/evm-semantics/.build/defn/haskell/evm-types.k)"), smtlib{}("_dotWS"), org'Stop'kframework'Stop'attributes'Stop'Location{}("Location(362,26,362,59)"), injective{}()]
```
Do you have any idea how to reproduce this in a unit test?

Reply:
Answered face to face.